x/bsonx/bsoncore/bsonbuilder.go

```diff
@@ -0,0 +1,362 @@
+	db.doc = AppendCodeWithScopeElement(db.doc, key, code, scope)
+	return db
+}
+
+// AppendCodeWithScope will append code and scope to a.arr
+func (a *ArrayBuilder) AppendCodeWithScope(code string, scope []byte) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendCodeWithScopeElement(a.arr, strconv.Itoa(a.key[last]), code, scope)
+	a.key[last]++
+	return a
+}
+
+// AppendTimestamp will append t and i to db.doc using provided key
+func (db *DocumentBuilder) AppendTimestamp(key string, t, i uint32) *DocumentBuilder {
+	db.doc = AppendTimestampElement(db.doc, key, t, i)
+	return db
+}
+
+// AppendTimestamp will append t and i to a.arr
+func (a *ArrayBuilder) AppendTimestamp(t, i uint32) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendTimestampElement(a.arr, strconv.Itoa(a.key[last]), t, i)
+	a.key[last]++
+	return a
+}
+
+// AppendInt64 will append i64 to dst using key to db.doc
+func (db *DocumentBuilder) AppendInt64(key string, i64 int64) *DocumentBuilder {
+	db.doc = AppendInt64Element(db.doc, key, i64)
+	return db
+}
+
+// AppendInt64 will append i64 to a.arr
+func (a *ArrayBuilder) AppendInt64(i64 int64) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendInt64Element(a.arr, strconv.Itoa(a.key[last]), i64)
+	a.key[last]++
+	return a
+}
+
+// AppendDecimal128 will append d128 to db.doc using provided key
+func (db *DocumentBuilder) AppendDecimal128(key string, d128 primitive.Decimal128) *DocumentBuilder {
+	db.doc = AppendDecimal128Element(db.doc, key, d128)
+	return db
+}
+
+// AppendDecimal128 will append d128 to a.arr
+func (a *ArrayBuilder) AppendDecimal128(d128 primitive.Decimal128) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendDecimal128Element(a.arr, strconv.Itoa(a.key[last]), d128)
+	a.key[last]++
+	return a
+}
+
+// AppendMaxKey will append a max key element using key to db.doc
+func (db *DocumentBuilder) AppendMaxKey(key string) *DocumentBuilder {
+	db.doc = AppendMaxKeyElement(db.doc, key)
+	return db
+}
+
+// AppendMaxKey will append a max key element to a.arr
+func (a *ArrayBuilder) AppendMaxKey(key string) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendMaxKeyElement(a.arr, strconv.Itoa(a.key[last]))
+	a.key[last]++
+	return a
+}
+
+// AppendMinKey will append a min key element using key to db.doc
+func (db *DocumentBuilder) AppendMinKey(key string) *DocumentBuilder {
+	db.doc = AppendMinKeyElement(db.doc, key)
+	return db
+}
+
+// AppendMinKey will append a min key element to a.arr
+func (a *ArrayBuilder) AppendMinKey(key string) *ArrayBuilder {
+	last := len(a.key) - 1
+	a.arr = AppendMinKeyElement(a.arr, strconv.Itoa(a.key[last]))
+	a.key[last]++
+	return a
+}
+
+// AppendInlineDocument starts building a document element with the provided key
+// After this document is completed, the user must call buildInlineDocument
+func (db *DocumentBuilder) AppendInlineDocument(key string) *DocumentBuilder {
```

Comment:
What do you think about the name `StartDocument` and `FinishDocument` for this and `BuildInlineDocument`, respectively?